### ISSUE TITLE
wai-extras: RequestLogger does not export enough to construct RequestLoggerSettings

### ISSUE THREAD

**Issue by jtdaugherty** [yesodweb/wai#365]
It appears that a `RequestLoggerSettings` cannot be constructed because the `RequestLogger` module doesn't export setters or constructors for the fields of the settings type.

**Comment by jtdaugherty** [comment]
(As of `3.0.6.1`.)

**Comment by DeathByTape** [comment]
This is not an issue. `RequestLoggerSettings` is an instance of `Default` which means you can do something like this (notice the `import Data.Default`):

```haskell
...
import Data.Default
import Network.Wai.Middleware.RequestLogger

defaultSettings :: RequestLoggerSettings
defaultSettings = def
...
```

or, if you want to override the default settings:

```haskell
...
import Data.Default
import Network.Wai.Middleware.RequestLogger

mySettings :: RequestLoggerSettings
mySettings = def {
  , outputFormat = Detailed False
    autoFlush = False
  }
-- Or any other settings you may wish to set. See doc for more info.
...
```

You can see the docs here: https://hackage.haskell.org/package/wai-extra-3.0.6.1/docs/Network-Wai-Middleware-RequestLogger.html

**Comment by jtdaugherty** [comment]
I see the source of the confusion: it is not documented that `outputFormat` and `autoFlush` are fields of `RequestLoggerSettings`, so it is not obvious that they can be used in the way you described.  Just judging by the types alone, they appear to be getters only.  Can you please update the haddock to indicate that this is possible?

**Comment by DeathByTape** [comment]
I have submitted the PR at https://github.com/yesodweb/wai/pull/366.

It is up to @snoyberg, though, if he wants to pull this in (or make alterations to the text).

**Comment by snoyberg** [comment]
PR merged, thank you for addressing that!

**Comment by jtdaugherty** [comment]
Thanks for the PR, @DeathByTape.  To be clear, the confusion was not about the `Default` instance's availability or purpose.  The PR does mention that, but I don't think it goes far enough to make it clear that "relevant fields" are literally *record fields* that can be used to update the default settings value in the way we've been discussing.  The way the haddock documentation gets generated hides this, so we need to do more to make it clear (to me, an example in the docs would be ideal).  I would be happy to submit a pull request of my own if you like.

Alternatively, another way to clear this up would be to simply export the constructor for the settings type.  If exported, it would make it completely clear in the haddock that the record fields are what they are.  I'm aware that often hiding constructors is a reasonable thing to do, but it doesn't seem like it buys much in this case and the resulting confusion has to be resolved with extra documentation.

**Comment by DeathByTape** [comment]
Ah, I see. Yes, I misunderstood. Sorry about that @jtdaugherty! 

Submitting your own PR is just as good as if I submitted one :) so that sounds like a good idea (read: I'm just some guy who recently started looking for ways to contribute to this project) ;)

In any case, I'm glad we cleared up the initial confusion. I would suggest you submit the PR with whatever you find to be the appropriate solution. It seems snoyberg is always pretty responsive at reviewing these things!
